Data/Primitive/Array.hs

```diff
@@ -591,6 +595,49 @@ arrayFromListN n l =
 arrayFromList :: [a] -> Array a
 arrayFromList l = arrayFromListN (length l) l
 
+filterArray :: forall a. (a -> Bool) -> Array a -> Array a
+filterArray f arr = runArray $
+  newBitArray s >>= check 0 0
+  where
+    s = sizeofArray arr
+    check :: Int -> Int -> MutableBitArray s -> ST s (MutableArray s a)
+    check i count ba
+      | i /= s
+      = do
+          v <- indexArrayM arr i
+          if f v
+            then setBitArray ba i >> check (i + 1) (count + 1) ba
+            else check (i + 1) count ba
+      | otherwise
+      = do
+          mary <- newArray count (die "filterArray" "invalid")
+          fill 0 0 ba mary
+
+    -- This performs a few bit operations and a conditional
+    -- jump for every element of the original array. This is
+    -- not so great if most element are filtered out. We should
+    -- consider going word by word through the bit array and
+    -- using countTrailingZeroes. We could even choose
+    -- a different strategy for each word depending on its
+    -- popCount.
```

Comment:
Also: I wouldn't be surprised if `unordered-containers` had some code or ideas we could steal here.

Reply:
I agree that CTZ (or equivalently CLZ) is going to do best here. I'm suspect that changing strategies with `popCount` would be helpful because I cannot think of a strategy that performs better when most of the elements are preserved. However, when *all* of them corresponding to a word of bits are preserved (meaning: the word is equal to `maxBound :: Word`), we could use the functions that copy a slice of the array instead. I guess we could also do this when `popCount` is really high instead of just when it's 64 (or 32 depending on platform).